SIGTRAP in WebCore::RenderFlexibleBox::marginBoxAscentForChild

pratiksha-99 opened the pull request:
#### 6f3345017b9769e62019a4910e8376d7b8dba03b
<pre>
SIGTRAP in WebCore::RenderFlexibleBox::marginBoxAscentForChild
<a href="https://bugs.webkit.org/show_bug.cgi?id=281992">https://bugs.webkit.org/show_bug.cgi?id=281992</a>
<a href="https://rdar.apple.com/137178953">rdar://137178953</a>

Reviewed by Alan Baradlay.

Changed to use a combination of min and max to avoid any invalid range for the flexItem.

* LayoutTests/fast/rendering/render-flexiblebox-margin-crash-expected.txt: Added.
* LayoutTests/fast/rendering/render-flexiblebox-margin-crash.html: Added.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::marginBoxAscentForFlexItem):

Canonical link: <a href="https://commits.webkit.org/285786@main">https://commits.webkit.org/285786@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/211bb66c9a6df3bc82789c8b5aa6421b69261d1b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73676 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53105 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26487 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77985 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24914 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75791 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62238 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/890 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57941 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16325 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76743 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48086 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63391 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38343 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44761 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20878 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23247 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66428 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21226 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79565 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/993 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/459 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66294 "Found 1 new test failure: http/tests/local/blob/navigate-blob.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1135 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63401 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65573 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16229 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9432 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7613 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/957 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3707 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/986 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/973 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/992 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1511,7 +1511,7 @@ LayoutUnit RenderFlexibleBox::marginBoxAscentForFlexItem(const RenderBox& flexIt
 
     if (isHorizontalFlow ? flexItem.isScrollContainerY() : flexItem.isScrollContainerX())
-        return std::clamp(*ascent, 0_lu, crossAxisExtentForFlexItem(flexItem)) + flowAwareMarginBeforeForFlexItem(flexItem);
+        return std::max(0_lu, crossAxisExtentForFlexItem(flexItem)) + flowAwareMarginBeforeForFlexItem(flexItem);
```
Oh no, that's not what I was suggesting. What I was suggesting is this:
```
std::clamp(*ascent, 0_lu, std::max(0_lu, crossAxisExtentForFlexItem(flexItem))) + flowAwareMarginBeforeForFlexItem(flexItem);
```

pratiksha-99 replied:
oh I get it now. Let me re-test.

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1511,7 +1511,7 @@ LayoutUnit RenderFlexibleBox::marginBoxAscentForFlexItem(const RenderBox& flexIt
 
     if (isHorizontalFlow ? flexItem.isScrollContainerY() : flexItem.isScrollContainerX())
-        return std::clamp(*ascent, 0_lu, crossAxisExtentForFlexItem(flexItem)) + flowAwareMarginBeforeForFlexItem(flexItem);
+        return std::clamp(*ascent, 0_lu, std::max(0_lu, crossAxisExtentForFlexItem(flexItem))) + flowAwareMarginBeforeForFlexItem(flexItem);
```
I would instead replace clamp with min/max.

pratiksha-99 replied:
I had that earlier but @rniwa suggested this is simpler.

alanbaradlay replied:
it may look simpler but I think the clamp usage here is incorrect as we can't guarantee a proper range.  We normally (and historically) use min/max for such cases as opposed to massaging the range (also this change can be seen as more error prone as now we've got to change the range value at 2 places if e.g. for whatever reason range is supposed to start at 1 instead of 0)

pratiksha-99 replied:
Thanks Alan for the explanation. I have updated the PR.

alanbaradlay commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1511,7 +1511,7 @@ LayoutUnit RenderFlexibleBox::marginBoxAscentForFlexItem(const RenderBox& flexIt
 
     if (isHorizontalFlow ? flexItem.isScrollContainerY() : flexItem.isScrollContainerX())
-        return std::clamp(*ascent, 0_lu, crossAxisExtentForFlexItem(flexItem)) + flowAwareMarginBeforeForFlexItem(flexItem);
+        return std::max(0_lu, crossAxisExtentForFlexItem(flexItem)) + flowAwareMarginBeforeForFlexItem(flexItem);
```
what happens now when *ascent is within the range?

pratiksha-99 replied:
It doesn't crash anymore and the test case passes.

alanbaradlay replied:
I think *ascent value now gets completely ignored.

pratiksha-99 replied:
Maybe thats why some other flexbox [tests](https://ews-build.s3-us-west-2.amazonaws.com/macOS-Ventura-Release-WK2-Tests-EWS/20350ad7-21124/results.html) are now failing.

alanbaradlay replied:
correct. clamp is supposed be replaced with the combination of min and max.

pratiksha-99 committed:
SIGTRAP in WebCore::RenderFlexibleBox::marginBoxAscentForChild
https://bugs.webkit.org/show_bug.cgi?id=281992
rdar://137178953

Reviewed by Alan Baradlay.

Changed to use a combination of min and max to avoid any invalid range for the flexItem.

* LayoutTests/fast/rendering/render-flexiblebox-margin-crash-expected.txt: Added.
* LayoutTests/fast/rendering/render-flexiblebox-margin-crash.html: Added.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::marginBoxAscentForFlexItem):

Canonical link: https://commits.webkit.org/285786@main
